DKMS: Support build and install for multiple kernel sources

frejanordsiek opened the pull request:
Currently, the `dkms.conf`

* Requires that `autogen.sh` be run on the directory first (prevents just giving the source tarball to dkms directly with `dkms ldtarball TARBALL`).
* Attempts to autodetect the kernel rather than use what DKMS indicates the kernel should be (build would not work well if more than one kernel is installed)
* Installs the library and headers even though only the module should be installed.

The `dkms.conf` file was changed to

* Run `autogen.sh` as the `PRE_BUILD` step
* Set the kernel directory and version with the information from DKMS
* Not run `make install` when building (means only the module will get installed after DKMS finds it). This also means that no install path is baked in.
* Removes comments about stuff the `dkms.conf` doesn't actually do

The one catch is that with these changes, the library and headers must be installed separately.

With this, one could download the source tarball and immediately do

```bash
dkms ldtarball xpmem-VERSION.tgz
dkms install xpmem/VERSION --all
```

and have the module be built for all kernels.

tvegas1 commented on dkms.conf:
```diff
@@ -1,19 +1,18 @@
 DEST_MODULE_LOCATION[0]="/kernel/../updates/"
 # how to build it
-MAKE[0]="./configure --prefix=/opt/xpmem; make clean; make install"
+MAKE[0]="./configure --prefix=/opt/xpmem --with-kerneldir=$kernel_source_dir --with-kernelvers=$kernelver ; make"
```
I understand that dkms now needs to own the .ko installation, hence the `make install` removal.

The userspace lib `libxpmem.so` and headers built with same `make` invocation can be installed with `make -C lib install`/`make -C include install`.

Is there a way to have dkms perform that installation? Maybe adding it to line 15, or having it triggered with separate dkms invocation/parameter?

yosefe replied:
AFAIK dkms should not install userspace stuff, so it should be ok

frejanordsiek replied:
That was the intention here -- separation of concerns with DKMS only handling the kernel module and installing the library and header being a completely separate thing to do (just like there are two RPM spec files, one for the library and one for the kernel module)

tvegas1 commented on dkms.conf:
```diff
@@ -1,19 +1,18 @@
 DEST_MODULE_LOCATION[0]="/kernel/../updates/"
 # how to build it
-MAKE[0]="./configure --prefix=/opt/xpmem; make clean; make install"
+MAKE[0]="./configure --prefix=/opt/xpmem --with-kerneldir=$kernel_source_dir --with-kernelvers=$kernelver ; make"
```
Shall we remove altogether if we don't install anything (but see previous comment)?

frejanordsiek replied:
I left the `--prefix` thing in there to make this the smallest change, and also least surprise if some bug in autotools causes the library to install when running just `make` (would put it where people usually put it, so minimized harm). I suppose some really random name in `/tmp` would also be suitable, or some place that would be impossible to write to like `/dev/null`.

By the way, I am working on some other changes (future PR) which would allow disabling the build of the library and installing headers entirely so that DKMS here would build the module and not waste any time building the unneeded library.

tvegas1 replied:
ok for me, but i think it's ok to remove prefix if we don't install

frejanordsiek replied:
It is removed now.

tvegas1 commented on dkms.conf:
```diff
@@ -1,19 +1,18 @@
-# install udev rule to control /dev file permissions
-
-
 # DKMS module name and version
```
Is it feasible to add similar steps below to CI under one of the existing dkms enabled docker as another `- bash section` around line below?
https://github.com/openucx/xpmem/blob/49d3e677fbd2983b1706957c27d384039ae68402/ci/azure-pipelines.yml#L116

```
./autogen.sh && ./configure --kernel... && make dist
dkms ldtarball xpmem-2.7.tgz
dkms install xpmem/2.7 --all
```

tvegas1 replied:
@Alexey-Rivkin

yosefe replied:
seems ./autogen.sh is not needed since it's added as PRE_BUILD step

tvegas1 replied:
if using `make dist` to create input tarball for dkms, autogen.sh+configure are needed to create the makefile.

frejanordsiek replied:
The idea was that you could download the source tarball directly from github or clone the repo and go to doing `make dist`, and then feed that tarball directly to dkms with no additional steps. But there is an important subtlety will I will discuss in a comment on the main chain I will soon make (e.g. the one downside is that autotools becomes a runtime dependency when using dkms, not allowing `autogen.sh` to be run once and only once).

tvegas1 replied:
trying to add dkms commands to CI using the git commit of the PR: without relying on `make dist` that ends up with a `./configure` file in the tarball, it's also possible to use on a clean repo `git archive --format=tar.gz -o xpmem-dkms.tar.gz --prefix xpmem-dkms/ master`, that can then be passed as input to dkms.

frejanordsiek replied:
I've changed the last commit to make it so it works regardless of whether `autogen.sh` was run or not before giving the tarball to dkms. `make dist` just has the advantage of removing autotools as a runtime dependency (good for any distro that might have this package)

tvegas1 reviewed (commented):
Maybe rename PR/Commits to `DKMS: Support build and install for multiple kernel sources`.

frejanordsiek commented:
> Maybe rename PR/Commits to `DKMS: Support build and install for multiple kernel sources`.

The PR has been renamed. The first commit was renamed to this as well. The second which added running `autogen.sh` to the `PRE_BUILD` step was renamed to `DKMS: Run autogen.sh as a PRE_BUILD step`


Now, I've thought of one major issue that must be discussed. By having the `PRE_BUILD="autogen.sh"`, autotools becomes a dependency to build and install the kernel module for every new kernel or when the module is rebuilt. There is no way, without a patch, to simply run `autogen.sh` once and then dkms doesn't have to run it each time (remove from `PRE_BUILD`) making it only a build dependency rather than also a runtime dependency.

I think this is a problem worth solving. My approach would be to add another script that be used by the `PRE_BUILD` step that would run `autogen.sh` only if any the `configure` script or any `Makefile.in` are missing (and any other files produced by autotools). This means that one could run `autogen.sh` once, give the result to dkms, and never worry about it again or give dkms the whole thing and `autogen.sh` be run each time. Thoughts on this?

tvegas1 commented:
> autotools). This means that one could run `autogen.sh` once, give the result to dkms, and never worry about it again or give dkms the whole thing and `autogen.sh` be run each time. Thoughts on this?

If we think adding autotools dependency is an issue but we still want to support passing tarball, maybe we can try to support both:
- `PRE_BUILD="test -f configure || ./autogen.sh"`

Commit:
DKMS: Support build and install for multiple kernel sources

frejanordsiek commented:
After the feedback, I've done some changes and another force push. The commits are slightly different and the last one was renamed. I will go through the changes.

1. No longer removing the comments about module loading and udev rules at the top. I realize now those are to guide any packager looking at the file about other steps they may want to consider doing.
2. `--prefix=/opt/xpmem` is no longer given to `configure` in the `dkms.conf` since `make install` isn't being used anymore.
3. The `PRE_BUILD` step was changed completely and a new script `dkms_prebuild.sh` added. See the reply below for more on that

> > autotools). This means that one could run `autogen.sh` once, give the result to dkms, and never worry about it again or give dkms the whole thing and `autogen.sh` be run each time. Thoughts on this?
> 
> If we think adding autotools dependency is an issue but we still want to support passing tarball, maybe we can try to support both:
> 
>     * `PRE_BUILD="test -f configure || ./autogen.sh"`

That is way simpler than I had in mind, and is sufficient. So, that is the code I have added, only changing the test from `-f` to `-x` (file must exist and be executable). Now, `PRE_BUILD` does not support shell commands and must instead be given a script to run. So, I just created a `dkms_prebuild.sh` script that contains just that, and put the script in `PRE_BUILD`.

tvegas1 commented on dkms_prebuild.sh:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env sh
+
+test -x configure || ./autogen.sh
```
In that case we can simply enhance `./autogen.sh` with something similar:
```diff
diff --git a/autogen.sh b/autogen.sh
index 2b47624..5f729c4 100755
--- a/autogen.sh
+++ b/autogen.sh
@@ -1,3 +1,17 @@
 #!/bin/sh

+if [ "$#" -gt 0 ]; then
+    if [ "$1" != --optional ]; then
+        echo "Usage: $0 [--optional]"
+        echo
+        echo "Options:"
+        echo "  --optional    Don't regenerate if already done"
+        exit 1
+    fi
+
+    if [ -x ./configure ]; then
+        exit 0
+    fi
+fi
+
 autoreconf -ivf || exit 1

```

frejanordsiek replied:
That is a cool thing to add to `autogen.sh`. However, it is still necessary to have `dkms_prebuild.sh` because `PRE_BUILD` only supports paths to a script, no arguments. But, with that addition, it would just become `./autogen.sh --optional` in `dkms_prebuild.sh`. This overall increases the complexity some. Thoughts on if that is worth it?

tvegas1 replied:
then maybe in that case, using your `dkms_prebuild.sh` with `test -x configure || ./autogen.sh` is enough

tvegas1 commented on dkms.conf:
```diff
@@ -6,14 +6,19 @@
 
+# Need to check if the configure script exists and run autogen.sh if it doesn't.
+# This has to be in a standalone script because DKMS does not support shell
+# commands in PRE_BUILD.
```
maybe add `or passing any argument`

frejanordsiek replied:
Done, though I used `nor` instead of `or`

tvegas1 commented on dkms_prebuild.sh:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env sh
```
maybe remove `/usr/bin/env` or add it to `autogen.sh` too?

frejanordsiek replied:
I added it to `autogen.sh` in the latest commit. This is important because with distros merging `/usr/bin` and `/bin`, it is on the table that `/bin` disappears and then `sh` is only at `/usr/bin/sh`. But `/usr/bin/env` is stable.

Commit:
DKMS: Add dkms_prebuild.sh script for PRE_BUILD step that runs autogen.sh if the configure script does not exist.

Commit:
DKMS: Changed the shebang in autogen.sh to use /usr/bin/env so that it will still work on Linux distros where /bin is gone and instead sh is at /usr/bin/sh.

frejanordsiek commented:
Just did another force push. I added more to the comment in `dkms.conf` for `PRE_BUILD` explaining why the script has to be used and I changed the shebang of `autogen.sh` to `#!/usr/bin/env sh` since some distros are preparing to get rid of `/bin` (then, `sh` is only at `/usr/bin/sh`).

tvegas1 commented:
Is it feasible to add the dkms steps similar as below in CI as another - bash section around line below?
```
git archive --format=tar.gz -o xpmem-dkms.tar.gz --prefix xpmem-dkms/ master
dkms ldtarball xpmem-dkms.tgz
module=`dkms status | awk  'BEGIN { FS="," } /xpmem/ { print $1 }'`
dkms build "$module"
```

[xpmem/ci/azure-pipelines.yml](https://github.com/openucx/xpmem/blob/49d3e677fbd2983b1706957c27d384039ae68402/ci/azure-pipelines.yml#L116)

frejanordsiek commented:
> Is it feasible to add the dkms steps similar as below in CI as another - bash section around line below?
> 
> ```
> git archive --format=tar.gz -o xpmem-dkms.tar.gz --prefix xpmem-dkms/ master
> dkms ldtarball xpmem-dkms.tgz
> module=`dkms status | awk  'BEGIN { FS="," } /xpmem/ { print $1 }'`
> dkms build "$module"
> ```
> 
> [xpmem/ci/azure-pipelines.yml](https://github.com/openucx/xpmem/blob/49d3e677fbd2983b1706957c27d384039ae68402/ci/azure-pipelines.yml#L116)

Just added it. Assuming the dockerfiles under `ci/dockers` are ultimately what the containers are made from, then it is only possible on the Ubuntu containers since they have `dkms` installed while the CentOS one does not (would be easy to change of course).

tvegas1 commented:
> Just added it. Assuming the dockerfiles under `ci/dockers` are ultimately what the containers are made from, then it is only possible on the Ubuntu containers since they have `dkms` installed while the CentOS one does not (would be easy to change of course).

Regarding CI failure, seems we don't have git installed, maybe tar can replace (or add to docker file) and dkms seems to need sudo.

frejanordsiek commented:
I just added `sudo` to the `dkms` calls and added `git` to both docker files and added `dkms` to the CentOS one so it can be tested to (required adding `epel-release` as well).

frejanordsiek commented:
The CI tests won't pass till the containers are rebuilt. Would you prefer that I change the CI to drop the test on CentOS and make the test work without git, or is it OK to leave it like this till the new containers get built?

tvegas1 commented:
can you please make it work without git on dkms-capable images, since regenerating containers might be delayed?

frejanordsiek commented:
> can you please make it work without git on dkms-capable images, since regenerating containers might be delayed?

I have done this but I am having a very hard time to get the test to actually succeed.

tvegas1 commented:
/azp run

Commit:
DKMS: Added doing a dkms build in the azure-pipelines CI.

tvegas1 commented:
> I have done this but I am having a very hard time to get the test to actually succeed.

Problem is a bug in dkms itself (https://github.com/dell/dkms/commit/468025752343dbc8609704bf0915f52b9c5c50f2), we can add line below with comment as workaround, before using dkms:
```
sed -i  -e 's@if \[\[ ! $temp_dir_name/dkms_main_tree \]\]@if \[\[ ! -d $temp_dir_name/dkms_main_tree \]\]@g' /usr/sbin/dkms
```

tvegas1 commented:
@frejanordsiek, do you plan to further work on that PR, or should I prepare a derivative PR to merge the change?

frejanordsiek commented:
Sorry to drop out on you like that. Was very busy for a while but have time again.

I would like to fix it. There was an incompatibility of some sort or some mistake on my part on Ubuntu that I wasn't able to resolve several weeks ago. I had added some diagnostics to the tests to try to figure out the problem (was trying to check if all files were actually present).

If you would like, I can try to further debug it and fix the issue.

tvegas1 commented:
No worries at all, I fixed the issues in #62 which should be ready, feel free to try it on your side.
